numpy/core/einsumfunc.py

```diff
@@ -0,0 +1,718 @@
+
+        contract = _find_contraction(contract_inds, input_sets, output_set)
+        out_inds, input_sets, idx_removed, idx_contract = contract
+
+        cost = _compute_size_by_dict(idx_contract, dimension_dict)
+        if idx_removed:
+            cost *= 2
+        cost_list.append(cost)
+        scale_list.append(len(idx_contract))
+        size_list.append(_compute_size_by_dict(out_inds, dimension_dict))
+
+        tmp_inputs = []
+        for x in contract_inds:
+            tmp_inputs.append(input_list.pop(x))
+
+        # Last contraction
+        if (cnum - len(path)) == -1:
+            idx_result = output_subscript
+        else:
+            sort_result = [(dimension_dict[ind], ind) for ind in out_inds]
+            idx_result = "".join([x[1] for x in sorted(sort_result)])
+
+        input_list.append(idx_result)
+        einsum_str = ",".join(tmp_inputs) + "->" + idx_result
+        contraction = (contract_inds, False, einsum_str, input_list[:])
+        contraction_list.append(contraction)
+
+    opt_cost = sum(cost_list)
+
+    if einsum_call_arg:
+        return (operands, contraction_list)
+
+    # Return the path along with a nice string representation
+    overall_contraction = input_subscripts + "->" + output_subscript
+    header = ("scaling", "BLAS", "current", "remaining")
+
+    path_print  = "  Complete contraction:  %s\n" % overall_contraction
+    path_print += "         Naive scaling:  %2d\n" % len(indices)
+    path_print += "      Naive FLOP count:  %.3e\n" % naive_cost
+    path_print += "  Optimized FLOP count:  %.3e\n" % opt_cost
+    path_print += "   Theoretical speedup:  %3.3f\n" % (naive_cost / float(opt_cost))
+    path_print += "  Largest intermediate:  %.3e elements\n" % max(size_list) 
+    path_print += "-" * 80 + "\n"
+    path_print += "%6s %6s %24s %40s\n" % header
+    path_print += "-" * 80
+
+    for n, contraction in enumerate(contraction_list):
+        inds, gemm, einsum_str, remaining = contraction
+        remaining_str = ",".join(remaining) + "->" + output_subscript
+        path_run = (scale_list[n], gemm, einsum_str, remaining_str)
+        path_print += "\n%4d    %6s %24s %40s" % path_run
+
+    return (path, path_print)
+
+# Rewrite einsum to handle different cases
+def einsum(*operands, **kwargs):
+    """
+    Evaluates the Einstein summation convention based on the operands,
+    differs from einsum by utilizing intermediate arrays to
+    reduce overall computational time.
+
+    Produces results identical to that of the einsum function; however,
+    the contract function expands on the einsum function by building
+    intermediate arrays to reduce the computational scaling and utilizes
+    BLAS calls when possible.
+
+    Parameters
+    ----------
+    subscripts : str
+        Specifies the subscripts for summation.
+    *operands : list of array_like
+        These are the arrays for the operation.
+    tensordot : bool, optional (default: True)
```

Comment:
Not currently no; however, tensordot features can be added so that contractions like `abcd,cdef->abef` can utilize a BLAS call (GEMM in this case). The downside is tensordot always copies data and your np.dot protocol does extra copies as well for C-ordered arrays. There was never a decision if the speed vs memory overhead was worth it.

See [here](https://github.com/dgasmith/opt_einsum/blob/master/opt_einsum/blas.py) for the BLAS lines that would be added.